runengine-server.md

```diff
@@ -0,0 +1,96 @@
+interfere with one another. For instance, for a long running scan, you do want
+a user closing a window to kill the background thread controlling hardware
+abruptly. (We can argue about whether actually want the scan to stop as closing
+a window will leave you without a clear way to manually pause or stop the scan,
+but I think we can agree that we want to guarantee that  all objects are
+unstaged and a stop document is produced. Keeping the RunEngine experimental
+control and data publishing logic in a separate process we can greatly increase stability.
+
+#### Centralization of RunEngine Configuration
+
+While the ability to quickly stand-up a RunEngine and start executing a scan is
+an incredibly powerful feature of `bluesky`, often instruments have a dedicated
+set of callbacks that are expected to be present. Keeping track of this desired
+configuration on where to save data e.t.c, and then ensuring that these are
+kept in-sync between GUIs / IPython profiles / Jupyter Notebooks, is painful. A
+benefit of going to a RunEngine server approach is you only need to launch the
+server with the desired configuration, then each of these processes only needs
+to know the information needed to know the location of the server.
+
+#### Experimental Control
+
+Centralizing the execution of scans can be beneficial to ensure that procedures
+are not adversely affecting each other. The fact that the server will not run
+plans in parallel would prevent operator errors where calibration routines are
+accidentally requested during data collection for example. This could be
+especially useful for automated tuning tasks that you want to run but do want
+to interfere during another user requested operation. 
+
+#### Experimental Gateway
+
+EPICS does not lend itself well to security, that is the nature of a
+distributed system. However, by setting up a RunEngine server that can be
+reached by a user, you can allow a specific restricted set of operations on a
+beamline without needing to give them permission to EPICS infrastructure at
+all. This could be incredibly useful for beamlines that are operated remotely
+via a web interface for example.
+
+## Issues and Pull Requests
+
+- https://github.com/bluesky/bluesky/pull/479
+
+## Affected Repositories
+
+- ophyd
+- bluesky
+
+## Open Design Decisions and Challenges
+
+#### Device Proxies
+Fortunately, `Msg` objects are trivial to serialize and ship over a wire, and
+for the most part the simple string and number arguments inside are as well.
+The one exception is how to correctly pass devices from client to server. The
+objects with live EPICS connections have to live on the server, the client has
+to be aware of which devices they can use in scans and how to specify each one.
+
+- The server could load a fixed set of devices. A user could then simply use
+  the "name" of devices instead of passing in the object itself. The server
+  would then be responsible for keeping a registry for lookup. This would require
```

Comment:
A thought I found in my paper notes on this: This would be easy to implement as an entry in `RE.preprocessors` that does the string->`OphydObj` lookup. That's really an implementation detail and may not belong in this document; I just wanted to mention it so we don't forget about that.